quinn/src/connection.rs

```diff
@@ -584,139 +542,172 @@ impl Connection {
+        let mut conn = self.0.state.lock("set_max_concurrent_bi_streams");
         conn.inner.set_max_concurrent_streams(Dir::Bi, count);
         // May need to send MAX_STREAMS to make progress
         conn.wake();
     }
 }
 
-/// A stream of unidirectional QUIC streams initiated by a remote peer.
-///
-/// Incoming streams are *always* opened in the same order that the peer created them, but data can
-/// be delivered to open streams in any order. This allows meaning to be assigned to the sequence in
-/// which streams are opened. For example, a file transfer protocol might designate the first stream
-/// the client opens as a "control" stream, using all others for exchanging file data.
-///
-/// Processing streams in the order they're opened will produce head-of-line blocking. For best
-/// performance, an application should be prepared to fully process later streams before any data is
-/// received on earlier streams.
-#[derive(Debug)]
-pub struct IncomingUniStreams(ConnectionRef);
-
-impl IncomingUniStreams {
-    /// Fetch the next incoming unidirectional stream
-    pub async fn next(&mut self) -> Option<Result<RecvStream, ConnectionError>> {
-        poll_fn(move |cx| self.poll(cx)).await
+pin_project! {
+    /// Future produced by [`Connection::open_uni`]
+    pub struct OpenUni<'a> {
+        conn: &'a ConnectionRef,
+        #[pin]
+        notify: Notified<'a>,
     }
+}
 
-    fn poll(&mut self, cx: &mut Context) -> Poll<Option<Result<RecvStream, ConnectionError>>> {
-        let mut conn = self.0.lock("IncomingUniStreams::poll");
-        if let Some(x) = conn.inner.streams().accept(Dir::Uni) {
-            conn.wake(); // To send additional stream ID credit
-            mem::drop(conn); // Release the lock so clone can take it
-            Poll::Ready(Some(Ok(RecvStream::new(self.0.clone(), x, false))))
-        } else if let Some(ConnectionError::LocallyClosed) = conn.error {
-            Poll::Ready(None)
-        } else if let Some(ref e) = conn.error {
-            Poll::Ready(Some(Err(e.clone())))
-        } else {
-            conn.incoming_uni_streams_reader = Some(cx.waker().clone());
-            Poll::Pending
-        }
+impl Future for OpenUni<'_> {
+    type Output = Result<SendStream, ConnectionError>;
+    fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let (conn, id, is_0rtt) = ready!(poll_open(ctx, this.conn, this.notify, Dir::Uni))?;
+        Poll::Ready(Ok(SendStream::new(conn, id, is_0rtt)))
     }
 }
 
-#[cfg(feature = "futures-core")]
-impl futures_core::Stream for IncomingUniStreams {
-    type Item = Result<RecvStream, ConnectionError>;
-
-    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
-        self.poll(cx)
+pin_project! {
+    /// Future produced by [`Connection::open_bi`]
+    pub struct OpenBi<'a> {
+        conn: &'a ConnectionRef,
```

Comment:
This will make it harder to use. For example, I will have to make a self-ref struct, however in the previous version I didn't have to:

```
pub struct QuicMuxer {
    connection: quinn::Connection,
    incoming: quinn::IncomingBiStreams,
    outgoing: Option<quinn::OpenBi>,
}
```

Reply:
That is why I asked a month ago at https://github.com/libp2p/rust-libp2p/pull/2801#issuecomment-1229357955 which futures you wanted to be `'static`.

This form avoids bit of indirection and allocation, and is able to reuse an existing implementation of a complex synchronization structure, so I'd like to make it work. What's the motivation for storing an `OpenBi`? Could you drop in `Box::pin(async move { conn.open_bi().await })` instead (until TAIT stabilizes)?

Reply:
> What's the motivation for storing an OpenBi

We got a poll_ based API.

> Could you drop in

I will figure that out tomorrow.

Reply:
Well, no.

Box::pin gives `Pin<Box<dyn Future<Output = Result<(quinn::SendStream, quinn::RecvStream), quinn::ConnectionError>>>>`, which is not Send:

```
 `(dyn futures::Future<Output = Result<(quinn::send_stream::SendStream, quinn::recv_stream::RecvStream), quinn_proto::connection::ConnectionError>> + 'static)` cannot be sent between threads safely
```

Reply:
You need to add `+ Send` to the trait object type.

Reply:
Thank you.

Reply:
To be clear, does that thereby address your requirements?

Reply:
Yes

Reply:
Excellent, thanks for the feedback!